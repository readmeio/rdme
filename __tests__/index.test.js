const nock = require('nock');
const cli = require('../src');
const { version } = require('../package.json');
const conf = require('../src/lib/configstore');

describe('cli', () => {
  it('command not found', async () => {
    await expect(cli('notARealCommand')).rejects.toThrow('Command not found');
  });

  describe('--version', () => {
    it('should return version from package.json', async () => {
      await expect(cli(['--version'])).resolves.toBe(version);
    });

    it('should return version if the `-V` alias is supplied', async () => {
      await expect(cli(['-V'])).resolves.toBe(version);
    });

    it('should return version from package.json for help command', async () => {
      await expect(cli(['help', '--version'])).resolves.toBe(version);
    });

    // This is necessary because we use --version for other commands like `docs`
    it('should only return version if no command', async () => {
      await expect(cli(['no-such-command', '--version'])).rejects.toThrow(
        // This can be ignored as it's just going to be a command not found error
        'Command not found.'
      );
    });

    it('should not be returned if `--version` is being used on a subcommand', async () => {
      nock.disableNetConnect();

      await expect(cli(['docs:edit', 'getting-started', '--version', '1.0.0', '--key=abcdef'])).rejects.not.toThrow(
        // We're testing that the docs:edit command does NOT return an error about `--version` not
        // being here because if it throws that error, then that means that `--version` wasn't
        // passed in as expected.
        'No project version provided. Please use `--version`.'
      );
    });
  });

  describe('--help', () => {
    beforeEach(() => {
      // The `table-layout` dependency within `command-line-usage` has a quirk in the 1.x family where if it's being
      // run within a Jest environment it won't always use the same amount of columns for tables, resulting in cases
      // where Jest snapshots will randomly return back different results.
      //
      // This issue has been fixed in later versions of `table-layout` but `command-line-usage` is still targeting the
      // 1.x series.
      //
      // See https://github.com/75lb/table-layout/issues/9 for more information.
      process.stdout.columns = 100;
      process.stderr.columns = 100;
    });

    afterEach(() => {
      process.stdout.columns = undefined;
      process.stderr.columns = undefined;
    });

    it('should print help', async () => {
      await expect(cli(['--help'])).resolves.toContain('a utility for interacting with ReadMe');
    });

    it('should print help for the `-H` alias', async () => {
      await expect(cli(['-H'])).resolves.toContain('a utility for interacting with ReadMe');
    });

    it('should print usage for a given command', async () => {
      await expect(cli(['openapi', '--help'])).resolves.toMatchSnapshot();
    });

    it('should print usage for a given command if supplied as `help <command>`', async () => {
      await expect(cli(['help', 'openapi'])).resolves.toMatchSnapshot();
    });

    it('should not surface args that are designated as hidden', async () => {
      await expect(cli(['openapi', '--help'])).resolves.toMatchSnapshot();
    });

    it('should show related commands for a subcommands help menu', async () => {
      await expect(cli(['versions', '--help'])).resolves.toMatchSnapshot();
    });
  });

  describe('subcommands', () => {
    // docs:edit will make a backend connection
    beforeAll(() => nock.disableNetConnect());

    it('should load subcommands from the folder', async () => {
      await expect(cli(['docs:edit', 'getting-started', '--version=1.0.0', '--key=abcdef'])).rejects.not.toThrow(
        'Command not found.'
      );
    });
  });

  it('should not error with undefined cmd', async () => {
    await expect(cli([])).resolves.toContain('Upload OpenAPI/Swagger definitions');
  });

  it('should add stored apiKey to opts', async () => {
    expect.assertions(1);
    conf.set('apiKey', '123456');

    await expect(cli(['docs'])).rejects.toThrow('No folder provided. Usage `rdme docs <folder> [options]`.');
    conf.delete('apiKey');
  });

  it('should error with `rdme oas` arguments passed in', async () => {
    await expect(cli(['oas', 'endpoint'])).rejects.toThrow(/.*/);
  });
});
