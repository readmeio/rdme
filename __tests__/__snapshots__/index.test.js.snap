// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli --help should not surface args that are designated as hidden 1`] = `
"
Upload, or resync, your OpenAPI/Swagger definition to ReadMe.

Usage

  rdme openapi [file] [options] 

Options

  --key string                Project API key                                                       
  --id string                 Unique identifier for your API definition. Use this if you're re-     
                              uploading an existing API definition                                  
  --version string            Project version                                                       
  --workingDirectory string   Working directory (for usage with relative external references)       
  -h, --help                  Display this usage guide                                              

Related commands

  $ rdme validate   Validate your OpenAPI/Swagger definition. 
  $ rdme swagger    Alias for \`rdme openapi\`. [deprecated]    
"
`;

exports[`cli --help should print help 1`] = `
"
                  üìñ rdme

    a utlity for interacting with ReadMe
       .
       .\\\\                          /.
      ‚Äô  ‚Äò                        ‚Äò ‚Äò
      ( nn\\\\    .           .     /  )
      (nnnnn,.MM.          AM   .nn.
       .nnnndMM----_______--M.‚Äô‚Äô   /
       |nnn/nnnnnnnnnnnnnnnnn\\\\‚Äômmm/
       /nnnn...nnnnnnnnnnn...mmmmm\\\\
      /nn        ‚ÄòqnnnP‚Äô       ‚Äòmmm\\\\
      /n‚Äô   .XXX. \\\\nnn/   .XX.  \\\\mmb
      An   (XOXX)  nnn   (XOXX)  mm\\\\
     /nn   ‚ÄòXXXX‚Äô.~‚Äù~.   ‚ÄòXXX‚Äù‚Äô .mmmb
     dnnn.      (    )n.       .mmmmb
    .nnnnnn....n.\\\\ ./nnnnnnnnnmmmmmm\\\\
  (READnnnnnnnnnnn‚ÄôY‚ÄônnnnnnnnnnmmmmmmME)
  REinnnnnnnnnnnnnnnnnnnnnnnnnmmmmmmmAD/
 /MEEnnnnnnnnnnnnnnnnnnnnnnnnnmmmmmmm)E'.
 READnnnnnnn*‚Äô             ‚Äò*mmmmmmmm)MEE.
(READ|nnnn‚Äô    \\\\../  \\\\.../    ‚ÄòmmmmmM)EEE)
 READ(nn*‚Äô                      ‚Äòmmm.MEEE)
  ‚ÄòREADn‚Äô  \\\\._./  \\\\__./  \\\\.../     ‚ÄòMEE*‚Äô
       *                           /*

Usage

  rdme <command> [arguments] 

Options

  -h, --help       Display this usage guide      
  -v, --version    Show the current rdme version 

Administration

  $ rdme login    Login to a ReadMe project.                                       
  $ rdme logout   Logs the currently authenticated user out of ReadMe.             
  $ rdme whoami   Displays the current user and project authenticated with ReadMe. 

Upload OpenAPI/Swagger definitions

  $ rdme openapi    Upload, or resync, your OpenAPI/Swagger definition to ReadMe. 
  $ rdme validate   Validate your OpenAPI/Swagger definition.                     
  $ rdme swagger    Alias for \`rdme openapi\`. [deprecated]                        

Documentation

  $ rdme docs        Sync a folder of markdown files to your ReadMe project.             
  $ rdme docs:edit   Edit a single file from your ReadMe project without saving locally. 

Versions

  $ rdme versions          List versions available in your project or get a version by SemVer       
                           (https://semver.org/).                                                   
  $ rdme versions:create   Create a new version for your project.                                   
  $ rdme versions:update   Update an existing version for your project.                             
  $ rdme versions:delete   Delete a version associated with your ReadMe project.                    

Other useful commands

  $ rdme oas    Helpful OpenAPI generation tooling.              
  $ rdme open   Open your current ReadMe project in the browser. 

Run rdme help <command> for help with a specific command.

To get more help with ReadMe, check out our docs at https://docs.readme.com.
"
`;

exports[`cli --help should print help for the \`-H\` alias 1`] = `
"
                  üìñ rdme

    a utlity for interacting with ReadMe
       .
       .\\\\                          /.
      ‚Äô  ‚Äò                        ‚Äò ‚Äò
      ( nn\\\\    .           .     /  )
      (nnnnn,.MM.          AM   .nn.
       .nnnndMM----_______--M.‚Äô‚Äô   /
       |nnn/nnnnnnnnnnnnnnnnn\\\\‚Äômmm/
       /nnnn...nnnnnnnnnnn...mmmmm\\\\
      /nn        ‚ÄòqnnnP‚Äô       ‚Äòmmm\\\\
      /n‚Äô   .XXX. \\\\nnn/   .XX.  \\\\mmb
      An   (XOXX)  nnn   (XOXX)  mm\\\\
     /nn   ‚ÄòXXXX‚Äô.~‚Äù~.   ‚ÄòXXX‚Äù‚Äô .mmmb
     dnnn.      (    )n.       .mmmmb
    .nnnnnn....n.\\\\ ./nnnnnnnnnmmmmmm\\\\
  (READnnnnnnnnnnn‚ÄôY‚ÄônnnnnnnnnnmmmmmmME)
  REinnnnnnnnnnnnnnnnnnnnnnnnnmmmmmmmAD/
 /MEEnnnnnnnnnnnnnnnnnnnnnnnnnmmmmmmm)E'.
 READnnnnnnn*‚Äô             ‚Äò*mmmmmmmm)MEE.
(READ|nnnn‚Äô    \\\\../  \\\\.../    ‚ÄòmmmmmM)EEE)
 READ(nn*‚Äô                      ‚Äòmmm.MEEE)
  ‚ÄòREADn‚Äô  \\\\._./  \\\\__./  \\\\.../     ‚ÄòMEE*‚Äô
       *                           /*

Usage

  rdme <command> [arguments] 

Options

  -h, --help       Display this usage guide      
  -v, --version    Show the current rdme version 

Administration

  $ rdme login    Login to a ReadMe project.                                       
  $ rdme logout   Logs the currently authenticated user out of ReadMe.             
  $ rdme whoami   Displays the current user and project authenticated with ReadMe. 

Upload OpenAPI/Swagger definitions

  $ rdme openapi    Upload, or resync, your OpenAPI/Swagger definition to ReadMe. 
  $ rdme validate   Validate your OpenAPI/Swagger definition.                     
  $ rdme swagger    Alias for \`rdme openapi\`. [deprecated]                        

Documentation

  $ rdme docs        Sync a folder of markdown files to your ReadMe project.             
  $ rdme docs:edit   Edit a single file from your ReadMe project without saving locally. 

Versions

  $ rdme versions          List versions available in your project or get a version by SemVer       
                           (https://semver.org/).                                                   
  $ rdme versions:create   Create a new version for your project.                                   
  $ rdme versions:update   Update an existing version for your project.                             
  $ rdme versions:delete   Delete a version associated with your ReadMe project.                    

Other useful commands

  $ rdme oas    Helpful OpenAPI generation tooling.              
  $ rdme open   Open your current ReadMe project in the browser. 

Run rdme help <command> for help with a specific command.

To get more help with ReadMe, check out our docs at https://docs.readme.com.
"
`;

exports[`cli --help should print usage for a given command 1`] = `
"
Upload, or resync, your OpenAPI/Swagger definition to ReadMe.

Usage

  rdme openapi [file] [options] 

Options

  --key string                Project API key                                                       
  --id string                 Unique identifier for your API definition. Use this if you're re-     
                              uploading an existing API definition                                  
  --version string            Project version                                                       
  --workingDirectory string   Working directory (for usage with relative external references)       
  -h, --help                  Display this usage guide                                              

Related commands

  $ rdme validate   Validate your OpenAPI/Swagger definition. 
  $ rdme swagger    Alias for \`rdme openapi\`. [deprecated]    
"
`;

exports[`cli --help should print usage for a given command if supplied as \`help <command>\` 1`] = `
"
Upload, or resync, your OpenAPI/Swagger definition to ReadMe.

Usage

  rdme openapi [file] [options] 

Options

  --key string                Project API key                                                       
  --id string                 Unique identifier for your API definition. Use this if you're re-     
                              uploading an existing API definition                                  
  --version string            Project version                                                       
  --workingDirectory string   Working directory (for usage with relative external references)       
  -h, --help                  Display this usage guide                                              

Related commands

  $ rdme validate   Validate your OpenAPI/Swagger definition. 
  $ rdme swagger    Alias for \`rdme openapi\`. [deprecated]    
"
`;

exports[`cli --help should show related commands for a subcommands help menu 1`] = `
"
List versions available in your project or get a version by SemVer (https://semver.org/).

Usage

  rdme versions [options] 

Options

  --key string       Project API key                                                 
  --version string   A specific project version to view                              
  --raw              Return raw output from the API instead of in a \\"pretty\\" format. 
  -h, --help         Display this usage guide                                        

Related commands

  $ rdme versions:create   Create a new version for your project.                
  $ rdme versions:update   Update an existing version for your project.          
  $ rdme versions:delete   Delete a version associated with your ReadMe project. 
"
`;
